cpp/src/arrow/compute/kernels/scalar_string.cc

```diff
@@ -1628,12 +1627,10 @@ const FunctionDoc utf8_slice_codeunits_doc(
 void AddSlice(FunctionRegistry* registry) {
   auto func = std::make_shared<ScalarFunction>("utf8_slice_codeunits", Arity::Unary(),
                                                &utf8_slice_codeunits_doc);
-  using t32 = SliceCodeunits<StringType>;
-  using t64 = SliceCodeunits<LargeStringType>;
-  DCHECK_OK(
-      func->AddKernel({utf8()}, utf8(), t32::Exec, SliceCodeunitsTransform::State::Init));
-  DCHECK_OK(func->AddKernel({large_utf8()}, large_utf8(), t64::Exec,
-                            SliceCodeunitsTransform::State::Init));
+  for (const auto& ty : StringTypes()) {
+    auto exec = GenerateTypeAgnosticVarBinary<SliceCodeunits>(ty);
```

Comment:
This function is named utf8_slice_codeunits, it shouldn't support binary types. "codeunits" doesn't make sense on a non-Unicode string and the implementation uses UTF-8 functions to count bytes; it'll give incorrect results on an arbitrary binary string.

Reply:
Only kernels for `StringTypes` are supported. Probably is the name of the GD `GenerateTypeAgnosticVarBinary` that creates confusion. `GenerateTypeAgnosticVarBinary` is not really `TypeAgnostic` but rather that output type is the same as input type.

Reply:
Ah, I see now. In that case, I don't think the generator really buys us that much here. But since the change is made already it's not a big deal.